### ISSUE TITLE
`unresolved-attribute` for circuitpython logging

### ISSUE THREAD

**Issue by alec-bike** [astral-sh/ty#2868]
### Summary

I use some CircuitPython libraries for an Adafruit board. Typing with ty has been fine with the use of `circuitpython-stubs` but recently I encountered a missing attribute error when using the `adafruit-circuitpython-logging` library:

```sh
uv add adafruit-circuitpython-logging
```

```py
import adafruit_logging as logging

logger = logging.getLogger("mylog")
logger.setLevel(logging.INFO)
                ^^^^^^^^^^^^
error[[unresolved-attribute](https://ty.dev/rules#unresolved-attribute)]: 
  Module `adafruit_logging` has no member `ERROR`
   --> src/bikelib/fcan/core.py:250:21
    |
248 |     # Setup diagnostic logging.
249 |     logger = logging.getLogger("fcan")
250 |     logger.setLevel(logging.ERROR)  # logging.INFO
    |                     ^^^^^^^^^^^^^
251 |     logger.info("Requesting closed-loop control...")
    |
info: rule `unresolved-attribute` is enabled by default
```

This passes `ty check`, and the code runs correctly on-device:

```py
logger.setLevel(20)
```

Notes:
- The attribute `INFO` is defined in `adafruit_logging.py`.
- I do not see this issue with the python `logging` library, only the CircuitPython variant.

### Version

ty 0.0.17 (8cec85718 2026-02-13)

**Comment by MeGaGiGaGon** [comment]
What's happening here is that `adafruit_logging` is doing some runtime trickery for the levels, and I'm not sure why:
https://github.com/adafruit/Adafruit_CircuitPython_Logging/blob/e4fac4691b86ac246c922c1c9382dd179bd5540a/adafruit_logging.py#L101C1-L113C32
```py
# The level module-global variables get created when loaded

LEVELS = [
    (00, "NOTSET"),
    (10, "DEBUG"),
    (20, "INFO"),
    (30, "WARNING"),
    (40, "ERROR"),
    (50, "CRITICAL"),
]

for __value, __name in LEVELS:
    globals()[__name] = __value
```

For reference, the built in `logging` module does it normally, which is why it works fine:
https://github.com/python/cpython/blob/e69501969be6c0dd44e7321f99e22f5499e98ce5/Lib/logging/__init__.py#L98-L105
```py
CRITICAL = 50
FATAL = CRITICAL
ERROR = 40
WARNING = 30
WARN = WARNING
INFO = 20
DEBUG = 10
NOTSET = 0
```

So because of the use of `globals` ty can't really do anything. (And the names being defined in `__all__` doesn't really help either, since ty could at most give a type of `Any` (though it currently doesn't and gives the error instead).)

**Comment by carljm** [comment]
Thanks @alec-bike for the report, and @MeGaGiGaGon for analysis!

For the case where a name is listed in `__all__` but not visibly present on the module, no other type checker currently silently accepts that. Mypy does the same as we do (normal doesn't-exist diagnostic). Pyright and Pyrefly offer custom diagnostics mentioning that the name is present in `__all__` but not defined. Pyrefly additionally errors on the `__all__` definition itself.

I think we should at least offer a custom diagnostic here, ideally, but not high priority.

**Comment by alec-bike** [comment]
Thanks for the feedback. I'll also draft a PR to fix this in CircuitPython (see https://github.com/adafruit/Adafruit_CircuitPython_Logging/issues/69).

**Comment by alec-bike** [comment]
PR [#70](https://github.com/adafruit/Adafruit_CircuitPython_Logging/pull/70) to fix the CircuitPython issue has been merged. I'll leave this ty issue open for reference.

**Comment by carljm** [comment]
We can keep this issue open to track the idea that we should offer a custom diagnostic in this case
